pkg/controllers/context/context.go

```diff
@@ -0,0 +1,103 @@
+package context
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+
+	"github.com/karmada-io/karmada/pkg/util/informermanager"
+	"github.com/karmada-io/karmada/pkg/util/objectwatcher"
+	"github.com/karmada-io/karmada/pkg/util/overridemanager"
+)
+
+// Options defines all the parameters required by our controllers.
+type Options struct {
+	// Controllers contains all controller names.
+	Controllers []string
+	// ClusterMonitorPeriod represents cluster-controller monitoring period, i.e. how often does
+	// cluster-controller check cluster health signal posted from cluster-status-controller.
+	// This value should be lower than ClusterMonitorGracePeriod.
+	ClusterMonitorPeriod metav1.Duration
+	// ClusterMonitorGracePeriod represents the grace period after last cluster health probe time.
+	// If it doesn't receive update for this amount of time, it will start posting
+	// "ClusterReady==ConditionUnknown".
+	ClusterMonitorGracePeriod metav1.Duration
+	// ClusterStartupGracePeriod specifies the grace period of allowing a cluster to be unresponsive during
+	// startup before marking it unhealthy.
+	ClusterStartupGracePeriod metav1.Duration
+	// ClusterStatusUpdateFrequency is the frequency that controller computes and report cluster status.
+	// It must work with ClusterMonitorGracePeriod.
+	ClusterStatusUpdateFrequency metav1.Duration
+	// ClusterLeaseDuration is a duration that candidates for a lease need to wait to force acquire it.
+	// This is measure against time of last observed lease RenewTime.
+	ClusterLeaseDuration metav1.Duration
+	// ClusterLeaseRenewIntervalFraction is a fraction coordinated with ClusterLeaseDuration that
+	// how long the current holder of a lease has last updated the lease.
+	ClusterLeaseRenewIntervalFraction float64
+	// ClusterCacheSyncTimeout is the timeout period waiting for cluster cache to sync.
+	ClusterCacheSyncTimeout metav1.Duration
+	// ClusterAPIQPS is the QPS to use while talking with cluster kube-apiserver.
+	ClusterAPIQPS float32
+	// ClusterAPIBurst is the burst to allow while talking with cluster kube-apiserver.
+	ClusterAPIBurst int
+	// SkippedPropagatingNamespaces is a list of namespaces that will be skipped for propagating.
+	SkippedPropagatingNamespaces []string
+	// ClusterName is the name of cluster.
+	ClusterName string
+}
+
+// Context defines the context object for controller.
+type Context struct {
+	Mgr                         controllerruntime.Manager
+	ObjectWatcher               objectwatcher.ObjectWatcher
+	Opts                        Options
+	StopChan                    <-chan struct{}
+	DynamicClientSet            dynamic.Interface
+	OverrideManager             overridemanager.OverrideManager
+	ControlPlaneInformerManager informermanager.SingleClusterInformerManager
+}
```

Comment:
The controllers running in `karmada-agent` and `karmada-controller-manager` might not always share the same `Context`. 
So, I'm thinking if it is worth abstracting a common `Context`.

Reply:
I have also considered this issue, but the agent is also a collection of several specific controllers, it's the same with controller-manager. With this abstraction, we can reuse part of the code

Reply:
> With this abstraction, we can reuse part of the code

Yes, That's the benefit. We also have to consider the maintenance effort that brings. Nothing right or wrong, just the tradeoff.

As far as I am concerned, if the benefits are not obvious I'd hesitate to do it.

Reply:
OK, then let's do it since you both insist.